Add util function for wildcard path matching

DanG100 opened the pull request:
* Enable less-strict path comparison for paths with wildcard keys

robshakir commented on util/gnmi.go:
```diff
@@ -53,12 +55,43 @@ func PathElemsEqual(a, b *gpb.PathElem) bool {
+
+// PathElemsEqualWildcard checks if two PathElem are the same allowing for wildcard keys.
+// If a.Key["foo"] == "*" and a.Key["foo"] == "bar" func returns true.
+// If a has no keys, it is to equal to b regardless of what keys b has and vice versa.
```
We probably should add something to the spec that defines this, but my expectation would be that `null` and `*` are the same thing - so that we say, for the list:

```yang
list foo { 
  key "one two";
}
```

* `/foo[one=*][two=*]`
* `/foo[one=*]`
* `/foo[two=*]`
* `/foo`

all mean the same thing (essentially, all entries in `foo`).

I didn't quite understand Wen's comment of "they are not on exactly the same keys" -- Wen, could you expand here?

Dan -- keys in YANG are mandatory to be specified, does your "unset key" here refer to in the supplied user wildcarded path?

wenovus replied:
It's irrelevant now. I believe I was saying how if the ordered pair of ("*", null, ...) between the two inputs don't match exactly, e.g. ("*", null) and (null, "*"), the current function would fail. But if we agree that `null` and `*` are the same thing, then they should be equal.

DanG100 replied:
thanks for clarification, i was talking about the user supplied path

DanG100 commented on util/gnmi.go:
```diff
@@ -53,12 +55,43 @@ func PathElemsEqual(a, b *gpb.PathElem) bool {
+}
+
+// PathElemsEqualWildcard checks if two PathElem are the same allowing for wildcard keys.
+// If a.Key["foo"] == "*" and a.Key["foo"] == "bar" func returns true.
```
outdated now

DanG100 commented:
Refactored this a bunch and simplified to follow the use case better. The goal of this func to check if a received gNMI path is valid based on the query path.

wenovus commented on util/gnmi.go:
```diff
@@ -43,6 +43,8 @@ func PathMatchesPrefix(path *gpb.Path, prefix []string) bool {
 }
 
 // PathElemsEqual replaces the proto.Equal() check for PathElems.
+// Note: Use PathElemsEquivalent for wildcards.
```
PathElemsEquivalent -> PathMatchesQuery

DanG100 replied:
moved  that comment to the PathMatchesPathElemPrefix

wenovus commented on util/gnmi_test.go:
```diff
@@ -526,6 +526,184 @@ func TestPathMatchesPathElemPrefix(t *testing.T) {
+				Key:  map[string]string{"seven": "eight"},
+			}},
+		},
+	}}
```
Another invalid test case: wildcard query but key is not present in return value.

DanG100 replied:
added

wenovus commented on util/gnmi.go:
```diff
@@ -92,6 +97,31 @@ func PathMatchesPathElemPrefix(path, prefix *gpb.Path) bool {
+	}
+	for i, queryElem := range query.Elem {
+		pathElem := path.Elem[i]
+		if queryElem == nil || pathElem == nil {
+			return false
+		}
```
Could you document this corner case behaviour?

DanG100 replied:
adde to godoc
